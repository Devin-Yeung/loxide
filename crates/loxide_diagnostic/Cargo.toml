[package]
name = "loxide_diagnostic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
miette = { workspace = true }

[dev-dependencies]
thiserror = { workspace = true }
insta = { workspace = true}