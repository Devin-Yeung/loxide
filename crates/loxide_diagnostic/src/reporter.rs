use miette::{GraphicalReportHandler, GraphicalTheme, Report};

#[derive(Debug)]
pub enum Style {
    Fancy,
    NoColor,
    #[cfg(test)]
    Test,
}

pub struct Reporter {
    handler: GraphicalReportHandler,
    diagnostics: Vec<Report>,
}

impl Reporter {
    pub fn new(style: Style) -> Self {
        let theme = match style {
            Style::Fancy => todo!(),
            Style::NoColor => GraphicalTheme::unicode_nocolor(),
            #[cfg(test)]
            Style::Test => GraphicalTheme::unicode_nocolor(),
        };

        Self {
            handler: GraphicalReportHandler::new_themed(theme),
            diagnostics: Vec::new(),
        }
    }

    pub fn push<T: Into<Report>>(&mut self, diagnostic: T) {
        self.diagnostics.push(diagnostic.into());
    }

    pub fn extend<T, I>(&mut self, diagnostic: I)
    where
        T: Into<Report>,
        I: IntoIterator<Item = T>,
    {
        self.diagnostics
            .extend(diagnostic.into_iter().map(Into::into));
    }

    pub fn report<W: std::fmt::Write>(&self, writer: &mut W) -> std::fmt::Result {
        for diagnostic in &self.diagnostics {
            self.handler.render_report(writer, diagnostic.as_ref())?
        }
        Ok(())
    }

    #[cfg(test)]
    pub fn report_to_string(&self) -> String {
        let mut buffer = String::new();
        self.report(&mut buffer).unwrap();
        buffer
    }
}

#[cfg(test)]
mod tests {
    use crate::reporter::Style;
    use miette::{Diagnostic, Report};
    use std::sync::Arc;
    use thiserror::Error;

    #[derive(Diagnostic, Debug, Error)]
    #[diagnostic()]
    pub enum ExampleError {
        #[error("Trailing space is not allowed")]
        NoTrailingSpace {
            #[label("remove this trailing space")]
            span: (usize, usize), // `(usize, usize)` is `Into<SourceSpan>`
        },
        #[error("Expected upper case letter")]
        ExpectedUpperCase {
            #[label("found lowercase letter")]
            span: (usize, usize), // `(usize, usize)` is `Into<SourceSpan>`
        },
    }

    #[test]
    fn unnamed_source() {
        let source = Arc::new(String::from("hello World! "));

        let diagnostics = vec![
            ExampleError::ExpectedUpperCase {
                span: (0 /* start offset */, 1 /* length */),
            },
            ExampleError::NoTrailingSpace { span: (12, 1) },
        ];

        let diagnostic = diagnostics
            .into_iter()
            .map(|d| Report::from(d).with_source_code(source.clone()))
            .collect::<Vec<_>>();

        let mut reporter = super::Reporter::new(Style::Test);
        reporter.extend(diagnostic);

        insta::assert_snapshot!(reporter.report_to_string());
    }
}
