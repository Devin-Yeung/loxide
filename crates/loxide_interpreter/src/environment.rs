use crate::error::RuntimeError;
use crate::value::Value;
use loxide_parser::ast::Identifier;
use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::{Arc, RwLock};

#[derive(Debug, Clone)]
struct Inner {
    values: HashMap<String, Value>,
    parent: Option<Arc<RwLock<Inner>>>,
}

impl PartialEq for Inner {
    fn eq(&self, other: &Self) -> bool {
        // check both values and parents
        self.values == other.values
            && match (&self.parent, &other.parent) {
                (Some(p1), Some(p2)) => p1.read().unwrap().eq(&p2.read().unwrap()),
                (None, None) => true,
                _ => false,
            }
    }
}

impl Inner {
    fn get(&self, ident: &Identifier) -> Result<Value, RuntimeError> {
        self.values.get(&ident.name).map_or_else(
            || {
                self.parent
                    .as_ref()
                    .map(|parent| parent.read().unwrap().get(ident))
                    .unwrap_or(Err(RuntimeError::UndefinedVariable(
                        ident.span,
                        ident.name.to_string(),
                    )))
            },
            |v| Ok(v.clone()),
        )
    }

    fn define(&mut self, ident: &Identifier, value: Value) {
        self.values.insert(ident.name.to_string(), value);
    }

    fn mutate(&mut self, ident: &Identifier, value: Value) -> Result<(), RuntimeError> {
        match self.values.get_mut(&ident.name) {
            Some(existed) => {
                *existed = value;
                Ok(())
            }
            None => match &self.parent {
                None => Err(RuntimeError::UndefinedVariable(
                    ident.span,
                    ident.name.to_string(),
                )),
                Some(env) => env.write().unwrap().mutate(ident, value),
            },
        }
    }

    /// freeze the current scope but not the parent scope
    fn freeze(&self) -> Inner {
        Inner {
            values: self.values.clone(),
            parent: self.parent.clone(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Environment {
    inner: Arc<RwLock<Inner>>,
}

impl PartialEq for Environment {
    fn eq(&self, other: &Self) -> bool {
        self.inner.read().unwrap().eq(&other.inner.read().unwrap())
    }
}

impl Environment {
    /// create a global scope
    pub fn global() -> Environment {
        Environment {
            inner: Arc::new(RwLock::new(Inner {
                values: HashMap::new(),
                parent: None,
            })),
        }
    }

    /// get the value of variable from the current scope,
    /// if current does not define this variable, search
    /// in the ancestor scope, if nothing found,
    /// return an UndefinedVariable runtime error
    pub fn get(&self, ident: &Identifier) -> Result<Value, RuntimeError> {
        self.inner.read().unwrap().get(ident)
    }

    /// define variable in current scope,
    /// if variable already defined in current scope,
    /// the previous value will be overwritten
    pub fn define(&mut self, ident: &Identifier, value: Value) {
        self.inner.write().unwrap().define(ident, value);
    }

    /// mutate variable in current scope,
    /// if current does not define this variable,
    /// mutate the variable in the ancestor scope,
    /// if nothing found in ancestor scope,
    /// return an UndefinedVariable runtime error
    pub fn mutate(&mut self, ident: &Identifier, value: Value) -> Result<(), RuntimeError> {
        self.inner.write().unwrap().mutate(ident, value)
    }

    /// create a descendant scope
    pub fn extend(&self) -> Environment {
        Environment {
            inner: Arc::new(RwLock::new(Inner {
                values: HashMap::new(),
                parent: Some(self.inner.clone()),
            })),
        }
    }

    /// freeze the current scope but not the parent scope
    pub fn freeze(&self) -> Environment {
        Environment {
            inner: Arc::new(RwLock::new(self.inner.read().unwrap().freeze())),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::environment::Environment;
    use crate::value::Value;
    use loxide_parser::ast::Identifier;
    use loxide_parser::token::Span;

    fn ident<S: AsRef<str>>(s: S) -> Identifier {
        Identifier {
            name: s.as_ref().to_string(),
            span: Span::new(0, 1),
        }
    }
    macro_rules! define {
        ($scope:expr, {$($name:expr => $val:expr),*$(,)?}) => {{
            $($scope.define(&ident($name), $val));*
        }};
    }

    macro_rules! check_scope {
        ($scope:expr, {$($name:expr => $val:expr),*$(,)?}) => {{
            $(assert_eq!($scope.get(&ident($name)).unwrap().as_number().unwrap(), $val);)*
        }};
    }

    macro_rules! mutate {
        ($scope:expr, {$($name:expr => $val:expr),*$(,)?}) => {{
            $($scope.mutate(&ident($name), $val).unwrap());*
        }};
    }

    #[test]
    fn nested() {
        let mut global = Environment::global();
        define!(global, {
            "a" => Value::Number(1.0),
        });

        let mut outer = global.extend();
        define!(outer, {
            "b" => Value::Number(2.0),
            "c" => Value::Number(0.0),
        });

        let mut inner = outer.extend();
        define!(inner, {
            "c" => Value::Number(3.0)
        });

        check_scope!(inner, {
            "a" => 1.0,
            "b" => 2.0,
            "c" => 3.0,
        });

        check_scope!(outer, {
            "c" => 0.0,
        });
    }

    #[test]
    fn update_outer_scope() {
        let mut global = Environment::global();
        define!(global, {
            "a" => Value::Number(1.0),
        });

        let mut inner = global.extend();
        mutate!(inner, {
            "a" => Value::Number(2.0)
        });
        // if inner scope does not define a var, but outer scope does
        // update will update the outer scope
        check_scope!(global, {
            "a" => 2.0,
        })
    }

    #[test]
    fn shadowing_dont_update_outer_scope() {
        let mut global = Environment::global();
        define!(global, {
            "a" => Value::Number(1.0),
        });

        let mut inner = global.extend();
        define!(inner, {
            "a" => Value::Number(88.0),
        });
        // update the inner scope
        mutate!(inner, {
            "a" => Value::Number(0.0)
        });

        check_scope!(global, {
            "a" => 1.0,
        });
        check_scope!(inner, {
            "a" => 0.0,
        });
    }

    #[test]
    fn inherit_outer_scope() {
        let mut global = Environment::global();
        define!(global, {
            "a" => Value::Number(1.0),
        });
        let inner = global.extend();
        check_scope!(inner, {
            "a" => 1.0,
        });
    }
}
