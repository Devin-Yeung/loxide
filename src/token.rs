#[derive(Debug, PartialEq)]
pub enum Literal {
    Identifier,
    String(std::string::String),
    Number(f64),
}

#[derive(Debug, PartialEq)]
pub enum Keyword {
    And,
    Class,
    Else,
    False,
    Fun,
    For,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,
}

#[derive(Debug, PartialEq)]
pub enum TokenType {
    // Single-character tokens.
    /// Opening parentheses `(`
    LeftParen,
    /// Closing parentheses `)`
    RightParen,
    /// Opening curly brace `{`
    LeftBrace,
    /// Closing curly brace `}`
    RightBrace,
    /// Comma `,`
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    // Comment
    Comment,
    EOF,
    // Literals.
    Literal(Literal),
    // Keywords.
    Keyword(Keyword),
}

#[derive(Debug)]
pub struct Token<'src> {
    pub(crate) ty: TokenType,
    pub(crate) lexeme: &'src str,
    pub(crate) span: Span,
}

/// The span of a token, line number indexed from 1,
/// column number indexed from 0, right exclusive.
#[derive(Debug, Copy, Clone)]
pub struct Span {
    pub(crate) line: usize,
    pub(crate) start: usize,
    pub(crate) end: usize,
}

impl<'src> Token<'src> {
    pub fn new(ty: TokenType, lexeme: &'src str, span: Span) -> Token<'src> {
        Token { ty, lexeme, span }
    }
}
