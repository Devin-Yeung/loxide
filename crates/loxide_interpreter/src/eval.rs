use crate::callable::Callable;
use crate::environment::Environment;
use crate::error::{PrivateRuntimeError, RuntimeError};
use crate::value::{Value, ValueKind};
use loxide_parser::ast::{
    AssignExpr, BinaryExpr, BinaryOperator, CallExpr, ConditionStmt, Expr, ExprKind, ForStmt,
    GroupedExpr, Identifier, Literal, ReturnStmt, Stmt, StmtKind, UnaryExpr, UnaryOperator,
    WhileStmt,
};
use loxide_testsuite::probe;
use std::io::Write;

pub trait Evaluable {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError>;
}

macro_rules! inner_or {
    ($val:expr, $ty:tt, $err:expr) => {{
        ::paste::paste! {
            if $val.[<is_$ty>]() {
                Ok($val.[<as_$ty>]().unwrap())
            } else {
                Err($err)
            }
        }
    }};
}

impl Evaluable for Stmt {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        self.kind.eval(env, stdout)
    }
}

impl Evaluable for StmtKind {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        match self {
            StmtKind::Expression(e) => e.eval(env, stdout),
            StmtKind::PrintStmt(e) => {
                let v = e.eval(env, stdout)?;
                let _ = probe!(&v, span => e.span().into());
                writeln!(stdout, "{v}").expect("Failed to write to stdout");
                Ok(Value::Void)
            }
            StmtKind::VarDeclaration(var, expr) => {
                let val = expr.eval(env, stdout)?;
                env.define(var, val);
                Ok(Value::Void)
            }
            StmtKind::FunDeclaration(decl) => {
                let callable = Callable::function(decl.clone(), env.extend());
                env.define(&decl.name, Value::Callable(callable));
                Ok(Value::Void)
            }
            StmtKind::Block(stmts) => {
                // TODO: find a way to test the correctness of this evaluation
                let mut scope = env.extend();
                for stmt in stmts {
                    stmt.eval(&mut scope, stdout)?;
                }
                Ok(Value::Void)
            }
            StmtKind::Condition(cond) => cond.eval(env, stdout),
            StmtKind::While(cond) => cond.eval(env, stdout),
            StmtKind::For(stmt) => stmt.eval(env, stdout),
            StmtKind::ReturnStmt(stmt) => stmt.eval(env, stdout),
            _ => unreachable!(),
        }
    }
}

impl Evaluable for ConditionStmt {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let mut env = env.extend();
        let guard = self.condition.eval(&mut env, stdout)?;
        match guard {
            Value::Boolean(b) => {
                if b {
                    self.then_branch.eval(&mut env, stdout)
                } else {
                    match &self.else_branch {
                        None => Ok(Value::Void),
                        Some(stmt) => stmt.eval(&mut env, stdout),
                    }
                }
            }
            _ => Err(RuntimeError::ExpectedBoolean(
                self.condition.span(),
                guard.kind(),
            )),
        }
    }
}

impl Evaluable for WhileStmt {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let cond = &self.condition;
        loop {
            let guard = cond.eval(env, stdout)?;
            match guard {
                Value::Boolean(b) => {
                    if b {
                        self.body.eval(env, stdout)?;
                    } else {
                        return Ok(Value::Void);
                    }
                }
                _ => return Err(RuntimeError::ExpectedBoolean(cond.span(), guard.kind())),
            }
        }
    }
}

impl Evaluable for ForStmt {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let mut env = env.extend();
        let _ = &self.initializer.eval(&mut env, stdout)?;
        match &self.condition {
            None => {
                // infinite loop
                loop {
                    self.body.eval(&mut env, stdout)?;
                    self.increment.eval(&mut env, stdout)?;
                }
            }
            Some(cond) => loop {
                let guard = cond.eval(&mut env, stdout)?;
                match guard {
                    Value::Boolean(b) => {
                        if b {
                            self.body.eval(&mut env, stdout)?;
                            self.increment.eval(&mut env, stdout)?;
                        } else {
                            return Ok(Value::Void);
                        }
                    }
                    _ => return Err(RuntimeError::ExpectedBoolean(cond.span(), guard.kind())),
                }
            },
        }
    }
}

impl Evaluable for ReturnStmt {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        self.value
            .as_ref()
            .map_or(Err(RuntimeError::ReturnValue(Value::Void)), |expr| {
                let value = expr.eval(env, stdout)?;
                Err(RuntimeError::ReturnValue(value))
            })
    }
}

impl Evaluable for Option<Box<Stmt>> {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        if let Some(stmt) = self {
            stmt.eval(env, stdout)
        } else {
            Ok(Value::Void)
        }
    }
}

impl Evaluable for Option<Box<Expr>> {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let val = match self {
            None => Value::Nil,
            Some(e) => e.eval(env, stdout)?,
        };
        Ok(val)
    }
}

impl Evaluable for Option<Expr> {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let val = match self {
            None => Value::Nil,
            Some(e) => e.eval(env, stdout)?,
        };
        Ok(val)
    }
}

impl Evaluable for Expr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        match &self.kind {
            ExprKind::Literal(l) => l.eval(env, stdout),
            ExprKind::Unary(u) => u.eval(env, stdout),
            ExprKind::Binary(b) => b.eval(env, stdout),
            ExprKind::Grouped(g) => g.eval(env, stdout),
            ExprKind::Variable(v) => v.eval(env, stdout),
            ExprKind::Assign(a) => a.eval(env, stdout),
            ExprKind::Call(c) => c.eval(env, stdout),
            _ => unreachable!(),
        }
    }
}

impl Evaluable for CallExpr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let callee = self.callee.eval(env, stdout)?;
        let args = self
            .args
            .iter()
            .map(|expr| expr.eval(env, stdout))
            .collect::<Result<Vec<Value>, RuntimeError>>()?;

        if let Value::Callable(callable) = callee {
            return match callable.call(args, env, stdout) {
                Ok(v) => Ok(v),
                Err(PrivateRuntimeError::BadArity { expected, found }) => {
                    Err(RuntimeError::BadArity {
                        callee_span: callable.params_span(),
                        caller_span: self.paren_token,
                        expected,
                        found,
                    })
                }
                Err(PrivateRuntimeError::Transparent(e)) => Err(e),
            };
        }

        Err(RuntimeError::CallOnNonCallable(
            self.callee.span(),
            callee.kind(),
        ))
    }
}

impl Evaluable for AssignExpr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let val = self.value.eval(env, stdout)?;
        env.mutate(&self.name, val.clone())?;
        Ok(val)
    }
}

impl Evaluable for BinaryExpr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let lhs = self.lhs.eval(env, stdout)?;
        let rhs = self.rhs.eval(env, stdout)?;
        let val = match &self.operator {
            op @ (BinaryOperator::EqualEqual | BinaryOperator::BangEq) => {
                return if lhs.kind() == rhs.kind()
                    && (lhs.kind() == ValueKind::Number
                        || lhs.kind() == ValueKind::Boolean
                        || lhs.kind() == ValueKind::String)
                {
                    match op {
                        BinaryOperator::EqualEqual => Ok(Value::Boolean(lhs == rhs)),
                        BinaryOperator::BangEq => Ok(Value::Boolean(lhs != rhs)),
                        _ => unreachable!(),
                    }
                } else {
                    Err(RuntimeError::IncompatibleBinaryOperand {
                        lhs_ty: lhs.kind(),
                        lhs_span: self.lhs.span(),
                        rhs_ty: rhs.kind(),
                        rhs_span: self.rhs.span(),
                    })
                }
            }
            BinaryOperator::GreaterEqual => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs >= rhs)
            }
            BinaryOperator::Greater => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs > rhs)
            }
            BinaryOperator::Less => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs < rhs)
            }
            BinaryOperator::LessEqual => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs <= rhs)
            }
            BinaryOperator::Minus => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs - rhs)
            }
            BinaryOperator::Plus => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs + rhs)
            }
            BinaryOperator::Slash => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs / rhs)
            }
            BinaryOperator::Star => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs * rhs)
            }
        };
        Ok(val)
    }
}

impl Evaluable for UnaryExpr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        let value = self.expr.eval(env, stdout)?;
        match self.operator {
            UnaryOperator::Minus => {
                let num = inner_or!(
                    value,
                    number,
                    RuntimeError::InvalidUnaryOperand(self.expr.span(), "number")
                )?;
                Ok(Value::Number(-num))
            }
            UnaryOperator::Bang => {
                let bool = inner_or!(
                    value,
                    boolean,
                    RuntimeError::InvalidUnaryOperand(self.expr.span(), "boolean")
                )?;
                Ok(Value::Boolean(!bool))
            }
        }
    }
}

impl Evaluable for GroupedExpr {
    fn eval(&self, env: &mut Environment, stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        self.expr.eval(env, stdout)
    }
}

impl Evaluable for Identifier {
    fn eval(&self, env: &mut Environment, _stdout: &mut impl Write) -> Result<Value, RuntimeError> {
        env.get(self)
    }
}

impl Evaluable for Literal {
    fn eval(
        &self,
        _env: &mut Environment,
        _stdout: &mut impl Write,
    ) -> Result<Value, RuntimeError> {
        let value = match *self {
            Literal::String(ref v) => Value::String(v.to_string()),
            Literal::Number(v) => Value::Number(v),
            Literal::Boolean(v) => Value::Boolean(v),
            Literal::Nil => Value::Nil,
        };
        Ok(value)
    }
}
