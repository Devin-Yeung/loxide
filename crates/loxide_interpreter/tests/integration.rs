use loxide_diagnostic::reporter::{Reporter, Style};
use loxide_interpreter::environment::Environment;
use loxide_interpreter::eval::Evaluable;
use loxide_parser::token::Span;
use miette::Report;
use std::ops::Range;
use std::sync::Arc;

struct LineSpan {
    /// the position of linebreaks
    linebreaks: Vec<usize>,
}

impl LineSpan {
    fn new<S: AsRef<str>>(src: S) -> Self {
        let linebreaks = src
            .as_ref()
            .char_indices()
            .filter_map(|(i, c)| if c == '\n' { Some(i) } else { None })
            .collect::<Vec<_>>();
        LineSpan { linebreaks }
    }

    /// Get the line span of the given span
    /// index start from 0, right exclusive
    fn get_line_span(&self, span: &Span) -> Range<usize> {
        let start = self
            .linebreaks
            .iter()
            .position(|index| span.start <= *index)
            .unwrap_or(self.linebreaks.len());
        let end = self
            .linebreaks
            .iter()
            .position(|index| span.end <= *index)
            .unwrap_or(self.linebreaks.len())
            + 1;
        start..end
    }
}

#[derive(PartialEq, Eq)]
struct Insertion {
    /// insertion after line #
    after: usize,
    /// non-descending id
    id: usize,
    content: String,
}

impl PartialOrd for Insertion {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Insertion {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.after.cmp(&other.after).then(self.id.cmp(&other.id))
    }
}

struct AnnotationBuilder {
    source: Arc<String>,
    reporter: Reporter,
    line_span: LineSpan,
    insertions: Vec<Insertion>,
    counter: usize,
}

impl AnnotationBuilder {
    fn new<S: AsRef<str>>(s: S) -> Self {
        let line_span = LineSpan::new(&s);
        let reporter = Reporter::new(Style::NoColor);
        let source = Arc::new(s.as_ref().to_string());
        AnnotationBuilder {
            source,
            reporter,
            line_span,
            insertions: Vec::new(),
            counter: 0,
        }
    }

    fn report_err<E: Into<Report>>(&mut self, err: E) {
        let report: Report = err.into();
        let report = report.with_source_code(self.source.clone());
        self.reporter.push(report)
    }

    /// annotate the evaluation result of a statement
    fn annotate(&mut self, content: String, span: Span) {
        let line_span = self.line_span.get_line_span(&span);
        self.insertions.push(Insertion {
            after: line_span.end - 1,
            id: self.counter,
            content,
        });
        self.counter += 1;
    }

    fn build(mut self) -> String {
        let mut source = self
            .source
            .as_str()
            .lines()
            .map(String::from)
            .collect::<Vec<_>>();
        self.insertions.sort();
        // insert in reverse order
        for insertion in self.insertions.into_iter().rev() {
            source.insert(insertion.after + 1, insertion.content);
        }
        if self.reporter.has_diagnostics() {
            source.push(format!(
                "\n>>>>> Error Section <<<<<\n{}",
                self.reporter.report_to_string()
            ));
        }
        #[cfg(target_os = "windows")]
        {
            source.join("\r\n")
        }
        #[cfg(not(target_os = "windows"))]
        {
            source.join("\n")
        }
    }
}

#[test]
fn test_line_span() {
    let src = "hello\nworld\n";
    let line_span = LineSpan::new(src);
    assert_eq!(line_span.get_line_span(&Span::new(0, 5)), 0..1);
    assert_eq!(line_span.get_line_span(&Span::new(6, 11)), 1..2);
    assert_eq!(line_span.get_line_span(&Span::new(1, 10)), 0..2);
    assert_eq!(line_span.get_line_span(&Span::new(4, 7)), 0..2);
}

fn annotated_eval<S: AsRef<str>>(src: S) -> String {
    let mut builder = AnnotationBuilder::new(&src);
    let mut parser = loxide_parser::parser::Parser::new(src.as_ref());
    let mut env = Environment::global();
    let (stmts, errs) = parser.parse();
    for err in errs {
        builder.report_err(err);
    }
    for stmt in stmts {
        let mut stdout = Vec::<u8>::new();
        let span = stmt.span();
        let _ = stmt
            .eval(&mut env, &mut stdout)
            .map_err(|e| builder.report_err(e));
        let content = String::from_utf8(stdout)
            .expect("Invalid utf8")
            .lines()
            .map(|line| format!("// => {}", line))
            .collect::<Vec<_>>()
            .join("\n");
        builder.annotate(content, span);
    }

    builder.build()
}

// #[test]
fn integration() {
    use walkdir::WalkDir;

    let files = WalkDir::new("fixture")
        .into_iter()
        .filter_map(Result::ok)
        .filter(|entry| entry.file_type().is_file())
        .collect::<Vec<_>>();

    for file in files {
        let content = std::fs::read_to_string(file.path())
            .unwrap_or_else(|_| panic!("Invalid file: {}", file.path().display()));

        // We don't support class yet
        let skip = vec!["class"];
        if skip.into_iter().any(|s| content.contains(s)) {
            continue;
        }

        let test_name = file.path().display().to_string().replace("\\", "/"); // I hate windows

        let skip = vec![
            // insane
            "benchmark",
            "limit",
            // panic
            "unexpected_character",
            // oop
            "field",
            // dead loop
            "number/decimal_point_at_eof",
            "number/trailing_dot",
            "string/unterminated",
            "scanning/numbers",
            // contains non-ascii
            "string/error_after_multiline",
            "for/fun_in_body",
            "string/literals",
            "unicode",
        ];
        if skip.into_iter().any(|s| test_name.contains(s)) {
            continue;
        }

        let test_name = test_name
            .strip_prefix("fixture/")
            .expect("fail to strip prefix")
            .strip_suffix(".lox")
            .expect("fail to strip suffix");
        println!("Running test: {}", test_name);

        let result = annotated_eval(&content);
        insta::with_settings!({
            snapshot_suffix => test_name,
            prepend_module_to_snapshot => false,
        }, {
            insta::assert_snapshot!(result)
        });
    }
}
