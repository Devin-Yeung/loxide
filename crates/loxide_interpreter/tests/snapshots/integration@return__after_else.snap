---
source: crates/loxide_interpreter/tests/integration.rs
expression: result
---
fun f() {
  if (false) "no"; else return "ok";
}


print f(); // expect: ok
// => str(ok)
